api/src/main/java/org/openmrs/api/db/hibernate/HibernateProgramWorkflowDAO.java

```diff
@@ -382,4 +393,86 @@ public ProgramWorkflow getWorkflowByUuid(String uuid) {
 		squery.setEntity("concept", concept);
 		return squery.list();
 	}
+        
+        @Override
+        public List<ProgramAttributeType> getAllProgramAttributeTypes() {
+            return sessionFactory.getCurrentSession().createCriteria(ProgramAttributeType.class).list();
+        }
+
+        @Override
+        public ProgramAttributeType getProgramAttributeType(Integer id) {
+            return (ProgramAttributeType) sessionFactory.getCurrentSession().get(ProgramAttributeType.class, id);
+        }
+
+        @Override
+        public ProgramAttributeType getProgramAttributeTypeByUuid(String uuid) {
+            return (ProgramAttributeType) sessionFactory.getCurrentSession().createCriteria(ProgramAttributeType.class).add(
+                    Restrictions.eq("uuid", uuid)).uniqueResult();
+        }
+
+        @Override
+        public ProgramAttributeType saveProgramAttributeType(ProgramAttributeType programAttributeType) {
+            sessionFactory.getCurrentSession().saveOrUpdate(programAttributeType);
+            return programAttributeType;
+        }
+
+        @Override
+        public PatientProgramAttribute getPatientProgramAttributeByUuid(String uuid) {
+            return (PatientProgramAttribute) sessionFactory.getCurrentSession().createCriteria(PatientProgramAttribute.class).add(Restrictions.eq("uuid", uuid)).uniqueResult();
+        }
+
+        @Override
+        public void purgeProgramAttributeType(ProgramAttributeType type) {
+            sessionFactory.getCurrentSession().delete(type);
+        }
+
```

Comment:
Did you intentionally leave out getPatientProgramByUuid, getPatientProgram, savePatientProgram, and getPatientPrograms?

Reply:
That's the way it is in the bahmni source

Reply:
Are you look at this commit? https://github.com/Bahmni/bahmni-core/commit/aeb460625de229de1581830c68d054b3caa40e94

Reply:
Yes, that's the one:
https://github.com/Bahmni/bahmni-core/commit/aeb460625de229de1581830c68d054b3caa40e94#diff-f30c98562e9f732e91631729c2e5f009R56

Reply:
Let me counter check

Reply:
`getPatientPrograms` is implemented the same exact way in the bahmni source as it is in the core, no need for change.

Reply:
same goes for  `getPatientProgramByUuid` the query does the same thing just written in a different way.

Reply:
as for `savePatientProgram` the only difference was 1 line:
`CustomDatatypeUtil.saveAttributesIfNecessary(patientProgram)` which I added.

Reply:
same goes for `getPatientProgram` the implementation is the same